src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/SortedTermsReader.java

```diff
@@ -0,0 +1,289 @@
+ * than just moving to the next term with {@link Cursor#advance()}.
+ * <p>
+ * For documentation of the underlying on-disk data structures, see the package documentation.
+ *
+ * @see SortedTermsWriter
+ * @see org.apache.cassandra.index.sai.disk.v1.sortedterms
+ */
+@ThreadSafe
+public class SortedTermsReader
+{
+    private final FileHandle termsData;
+    private final SortedTermsMeta meta;
+    private final FileHandle termsTrie;
+    private final LongArray.Factory blockOffsetsFactory;
+
+    /**
+     * Creates a new reader based on its data components.
+     * <p>
+     * It does not own the components, so you must close them separately after you're done with the reader.
+     * @param termsData handle to the file with a sequence of prefix-compressed blocks
+     *                  each storing a fixed number of terms
+     * @param termsDataBlockOffsets handle to the file containing an encoded sequence of the file offsets pointing to the blocks
+     * @param termsTrie handle to the file storing the trie with the term-to-point-id mapping
+     * @param meta metadata object created earlier by the writer
+     * @param blockOffsetsMeta metadata object for the block offsets
+     */
+    public SortedTermsReader(@Nonnull FileHandle termsData,
+                             @Nonnull FileHandle termsDataBlockOffsets,
+                             @Nonnull FileHandle termsTrie,
+                             @Nonnull SortedTermsMeta meta,
+                             @Nonnull NumericValuesMeta blockOffsetsMeta) throws IOException
+    {
+        this.termsData = termsData;
+        this.termsTrie = termsTrie;
+        try (IndexInput trieInput = IndexInputReader.create(termsTrie))
+        {
+            SAICodecUtils.validate(trieInput);
+        }
+        this.meta = meta;
+        this.blockOffsetsFactory = new MonotonicBlockPackedReader(termsDataBlockOffsets, blockOffsetsMeta);
+    }
+
+    /**
+     * Returns the point id (ordinal) of the target term or the next greater if no exact match found.
+     * If reached the end of the terms file, returns <code>Long.MAX_VALUE</code>.
+     * Complexity of this operation is O(log n).
+     *
+     * @param term target term to lookup
+     */
+    public long getPointId(@Nonnull ByteComparable term)
+    {
+        Preconditions.checkNotNull(term, "term null");
+
+        try (TrieRangeIterator reader = new TrieRangeIterator(termsTrie.instantiateRebufferer(),
+                                                              meta.trieFP,
+                                                              term,
+                                                              null,
+                                                              true,
+                                                              true))
+        {
+            final Iterator<Pair<ByteSource, Long>> iterator = reader.iterator();
+            return iterator.hasNext() ? iterator.next().right : Long.MAX_VALUE;
```

Comment:
nit: Do any tests hit the `return Long.MAX_VALUE` branch?

Reply:
I do not understand why it is returning `Long.MAX_VALUE` at this point. The user of this method `PostingListRangeIterator.getNextRowId` has the following test on the return value:
```
            if (targetRowID < 0)
            {
                return PostingList.END_OF_STREAM;
            }
```
This implies that it is expecting the end of stream marker to be `< 0` not `Long.MAX_VALUE`. I will investigate this and make sure that it is correctly tested.